File: packages/javascript/lib/array/arrUnique.ts
import { is2dArray } from '../verify/is2dArray.js';

export function arrUnique(array: any[]): any[] {
	if (is2dArray(array)) {
		const seen = new Set<string>();
		const result: any[] = [];

		for (let i = 0; i < array.length; i += 1) {
			const key = JSON.stringify(array[i]);

			// `undefined` and functions have no JSON representation, so they cannot be compared
			// this way. Keep them as-is rather than throwing on `JSON.parse(undefined)`.
			if (key === undefined) {
				result.push(array[i]);
			} else if (!seen.has(key)) {
				seen.add(key);
				result.push(array[i]);
			}
		}

		return result;
	}

	return [...new Set(array)];
}

File: packages/python/qsu/array/arrUnique.py
import json

from ..verify.is2dArray import is2dArray


def arrUnique(array: list) -> list:
	if is2dArray(array):
		seen = set()
		result = []

		for item in array:
			key = json.dumps(item, separators=(',', ':'))

			if key not in seen:
				seen.add(key)
				result.append(json.loads(key))

		return result

	seen = []
	result = []

	for item in array:
		if not any(existing is item or (type(existing) is type(item) and existing == item) for existing in seen):
			seen.append(item)
			result.append(item)

	return result
